lib/FusionInventory/Agent/Task/Inventory/Linux/Storages.pm

```diff
@@ -56,24 +65,29 @@ sub _getDevices {
         }
     }
 
-    # get serial & firmware numbers from hdparm, if available
-    if (_correctHdparmAvailable(%params)) {
-        foreach my $device (@devices) {
-            next if $device->{SERIALNUMBER} && $device->{FIRMWARE};
-            my $info = getHdparmInfo(
-                device => "/dev/" . $device->{NAME},
-                %params
+    # get missing fields using functions defined in %map
+    for my $device (@devices) {
+        # the hash keys are function references from %map
+        my %info;
+
+        OUTER: for my $field (keys %map) {
+            next unless (
+                ($field eq 'MANUFACTURER'
+                 && defined $device->{$field}
+                 && $device->{$field} eq 'ATA')
+                      ||
+                !$device->{$field}
             );
 
-            $device->{SERIALNUMBER} = $info->{serial}
-                if $info->{serial} && !$device->{SERIALNUMBER};
+            INNER: for my $sub (@{$map{$field}}) {
+                # get info once for each device
+                $info{$sub} = &$sub(device => '/dev/' . $device->{NAME}, %params) unless $info{$sub};
 
-            $device->{FIRMWARE} = $info->{firmware}
-                if $info->{firmware} && !$device->{FIRMWARE};
-
-            $device->{DESCRIPTION} = $info->{transport} if $info->{transport};
-            $device->{MODEL}       = $info->{model} if $info->{model};
-            $device->{WWN}         = $info->{wwn} if $info->{wwn};
+                if (defined $info{$sub}->{$field}) {
+                    $device->{$field} = $info{$sub}->{$field};
+                    next OUTER;
```

Comment:
In place of using OUTER & INNER, I think just replacing `next OUTER;` by `last;` is sufficient as we leave INNER and nothing is done before looping on OUTER.